fabtests/common/shared.c

```diff
@@ -3142,7 +3166,21 @@ int ft_check_buf(void *buf, size_t size)
 	int msg_index = 0;
 	size_t i;
 
-	recv_data = (char *)buf;
+	if (opts.iface != FI_HMEM_SYSTEM) {
+		recv_data = malloc(size);
+		if (!recv_data) {
+			free(recv_data);
+			return -FI_EIO;
+		}
+	
+		if (ft_hmem_copy_from(opts.iface, opts.device,
+				      recv_data, buf, size)) {
+			free(recv_data);
+			return -FI_EIO;
```

Comment:
Restructure the code in this function to have a common exit where the memory is freed, similar to above example.